Adjust GcpLayout JSON to latest format

ViliusS opened the pull request:
This is my first contribution to Log4J so I'm not 100% sure if I did everything right, but here it goes.

This PR adjusts GcpLayout JSON formatting to match what is expected by GCP.

First, it formats the log timestamp field to the correct format recognized by Fluent-Bit (component of Google Cloud Logging) and Google Ops Agent. I'm not sure how it worked before, probably it didn't, but Google Cloud Logging [expects timestampSecond + timestampNanos format](https://cloud.google.com/logging/docs/agent/logging/configuration#timestamp-processing). Ops Agent, which also can be used to feed the logs from external Log4J environment via Socket or RollingFile into GCP, [also expects the same format](https://cloud.google.com/logging/docs/agent/ops-agent/configuration#special-fields). More over, the same [timestamp format is always used in Spring GCP libraries](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/cabb7a59fa6d49d280dd311a7f59ec4d922300e7/spring-cloud-gcp-logging/src/main/java/com/google/cloud/spring/logging/StackdriverJsonLayout.java#L270-L278).

Secondly, severity field now must be prefixed with `logging.googleapis.com`. I'm not 100% sure regarding this, because Spring GCP libraries are still using just `severity` field name. However Ops Agent (link above) already uses `logging.googleapis.com/severity`, and simple `severity` name is only described in legacy logging agent which is deprecated.

Third, counter cannot be used for `logging.googleapis.com/insertId` as it is duplicated on different threads. I've chosen to use time-based UUID, but if that increases garbage collection issues it safely can be removed. GCP will generate their own IDs anyway.

And the last but not the least, I have renamed `exception`, `thread` and `logger` fields. I saw the PR discussion in https://github.com/apache/logging-log4j2/pull/543, however these fields are pretty standard when logging via [Logback's JSON layout](https://github.com/qos-ch/logback-contrib/blob/c68c2e5ddc0ed589b0f1e24bac7b9300a0ea5153/json/classic/src/main/java/ch/qos/logback/contrib/json/classic/JsonLayout.java#L105-L113) and then they are reused in Google's Spring GCP libraries. Sadly, I didn't find how to completely remove `exception` field from JSON output if it's empty. Looks like all empty fields are removed except for `pattern` type. I also found https://issues.apache.org/jira/browse/LOG4J2-3053 regarding that, but not sure about its state. This is not a big issue but is preferable for GCP.
By the way, Log4J docs are a little bit incorrect. If `message` field exist, having separate `exception` does nothing to catch the error in Google Error Reporting. According to [Google docs](https://cloud.google.com/error-reporting/docs/formatting-error-messages#format-log-entry), if `message` field exist it must contain stack trace and only that stack trace is used in the Error Reporting UI. I tested this extensively. But I left separate `exception` field if one wants to filter it in the logs.

All in all, this should provide much better JSON field mapping experience on GCP. If you think this PR is usable I then will go ahead and fix tests for timestamp processing.

ppkarwasz commented on log4j-layout-template-json/src/main/resources/GcpLayout.json:
```diff
@@ -49,25 +55,15 @@
     "field": "name"
   },
-  "_logger": {
+  "logger": {
```
Those field start intentionally with an underscore `_`, so they don't collide with payload field that Google might introduce in the future.

ViliusS replied:
They are already used by Google, just not documented. For example in their Spring GCP libraries https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/cabb7a59fa6d49d280dd311a7f59ec4d922300e7/spring-cloud-gcp-logging/src/main/java/com/google/cloud/spring/logging/StackdriverJsonLayout.java#L286-L287. My idea was to match the format Google already produces officially by their Java libraries as much as possible.

ppkarwasz replied:
Sounds good to me!

ppkarwasz reviewed (changes requested):
@ViliusS,

Thanks for you contribution.

Did you test it on an actual Google Cloud instance? I am not a GCP user, but some of the changes you propose differ from the documented log format.

ViliusS commented on log4j-layout-template-json/src/main/resources/GcpLayout.json:
```diff
@@ -1,13 +1,18 @@
-    "pattern": {
-      "format": "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
-      "timeZone": "UTC",
-      "locale": "en_US"
+    "epoch": {
+      "unit": "secs",
+      "rounded": true
     }
   },
-  "severity": {
+  "timestampNanos": {
+    "$resolver": "timestamp",
+    "epoch": {
+      "unit": "secs.nanos"
+    }
+  },
```
I specifically chosen to use `{"timestampSecond", "timestampNanos"}` format because:
a) It is common for all three Google application: Spring GCP libraries, Ops Agent, Legacy Logging Agent.
b) Google Ops Agent JSON processor is recursive, so if you have more than one level in JSON it is impossible to operate on that field later in Ops Agent pipeline and match correctly to the timestamp field. See https://github.com/GoogleCloudPlatform/ops-agent/issues/1920#issuecomment-2769268587.

ViliusS commented on log4j-layout-template-json/src/main/resources/GcpLayout.json:
```diff
@@ -1,13 +1,18 @@
+      "unit": "secs.nanos"
+    }
+  },
+  "logging.googleapis.com/severity": {
```
From explanation here it looks like this depends how you send the logs to Google Cloud Logging.
If Ops Agent is used, you should use `logging.googleapis.com/severity` OR use `severity` and adjust it later in Ops Agent pipeline as described in this example https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent/configuration#example_configuration_2.
If log is generated into the file and then send to Google Cloud Logging directly from Java, for example using Spring GCP libraries, then it must be `severity`.

Full explanation from Google support here https://github.com/GoogleCloudPlatform/ops-agent/issues/1920#issuecomment-2772380200 . I think I'm going to revert this to `severity` because there is a workaround for Ops Agent.

ViliusS replied:
Reverted in https://github.com/apache/logging-log4j2/commit/b19491c451333464442416eff462bbbd5750cd00

ViliusS commented on log4j-layout-template-json/src/main/resources/GcpLayout.json:
```diff
@@ -49,25 +55,15 @@
-      "$resolver": "pattern",
-      "pattern": "%xEx"
-    }
+  "exception": {
+    "$resolver": "pattern",
+    "pattern": "%xEx"
   },
```
I didn't find detailed reference for these fields, but that description in documentation is confusing a bit. What they mean is that they check `stack_trace` first, but they don't stop. They check all the fields present and then if `message` field exist they just take stack trace from `message` field ignoring all others.

At least this is how it worked for me during testing (I spent 2 hours trying various stack trace formats and fields in real GCP environment).

I also tried to use `exception` resolver but it produces just a stack trace without header and is not cought by Google Error Reporting. So in the end I left `pattern` resolver like it was originally.

ppkarwasz replied:
> I also tried to use `exception` resolver but it produces just a stack trace without header and is not cought by Google Error Reporting. So in the end I left `pattern` resolver like it was originally.

That is interesting, the snippet above produces **exactly** the same output as `Throwable.printStackTrace()`. @vy, any ideas?

My preferences for these field are in order:

1. `exception` resolver if it works.
2. `pattern` with a pattern of `%ex` (less computationally intensive)
3. what is currently proposed.

vy replied:
> I also tried to use `exception` resolver but it produces just a stack trace without header

@ViliusS, could you elaborate on this, please? What does _"a stack trace without header"_ mean? Can you provide a test that we can reproduce the problem on our side?

As we stated in JTL docs, falling back to Pattern Layout should be a last resort. In the case of exception rendering, ideally, one should never need to fallback to PL.

ViliusS replied:
It seems that I first tested `exception` resolver without `stringify` parameter. This was a probable reason why it was not caught by Google Error Reporting dashboard. I have now re-tested with `stringify` -> `true` and this is what I've got.

The look with `exception` resolver:
![image](https://github.com/user-attachments/assets/42a6a1ec-9162-4ea8-a06c-074f3427b1c2)

This is how it looks with `%eXe` pattern:
![image](https://github.com/user-attachments/assets/7a5e4512-c0ef-4369-8126-898045a40571)

This is the look on Error Reporting page (item 1 and 2 in the list):
![image](https://github.com/user-attachments/assets/a6dea5b5-f402-4a5a-a53c-dff6e1100fb6)

If you ask me I would still preferred message + exception format (item 3 in the screenshot above). At least this is how other libraries and Java software reports exceptions to Error Reporting. See the comparison:
![image](https://github.com/user-attachments/assets/21939493-e0c2-413b-8ca4-94cee48735df)

But to match this, we would need to remove `exception` field completely and rely only on `message` field, or have the same information duplicated to `exception` field, which I don't see the point.

Anyway, if you prefer one format or another, I could adjust the patch. Keep in mind though that current patch doesn't change anything regarding exception format. Current JSON layout used %eXe all along. I just changed field name.

ViliusS replied:
Just remembered one more issue why I have left `%xEx`. As `exception` resolver doesn't add full package name to the exception, all `ActiveMQIllegalState` exceptions, even from other packages but having different stacktrace, would be grouped into one error group. I haven't tested this with this exact case, but I know from my previous work with GCP and Java exceptions that full package name is very important to group errors appropriately.

ppkarwasz replied:
> As `exception` resolver doesn't add full package name to the exception, all `ActiveMQIllegalState` exceptions, even from other packages but having different stacktrace, would be grouped into one error group.

The `exception` resolver prints exactly the same text as `Throwable#printStackTrace`. What probably happens is that GCP recognizes that format and strips the package name. He doesn't recognize `%xEx`.

However, I would go with `exception` and ask GCP to store the FQCN of the exception.

ViliusS replied:
> > As `exception` resolver doesn't add full package name to the exception, all `ActiveMQIllegalState` exceptions, even from other packages but having different stacktrace, would be grouped into one error group.
> 
> The `exception` resolver prints exactly the same text as `Throwable#printStackTrace`. What probably happens is that GCP recognizes that format and strips the package name. He doesn't recognize `%xEx`.

Google Error Reporting nor Cloud Logging does that. They just process log messages for various indicators and print them as they are. I also get the same format in plaintext JSON file before sending it to GCP, so this is definitely not processed on their side.

Unless there is some kind of bug in my Log4J version (2.24.3) or Artemis app modifies internal exception format somehow.

ppkarwasz replied:
@ViliusS,

Do you have a JIRA account to report this in the [Artemis JIRA](https://issues.apache.org/jira/projects/ARTEMIS) or should we report it for you? I guess that the correct `toString` should be:

```
org.apache.activemq.artemis.api.core.ActiveMQIllegalStateException: [errorType=ILLEGAL_STATE message=]
```

Probably Artemis itself is relying on `%xEx`, so they'll have a little surprise when they upgrade to `2.25.0`, since `%ex` will be the default.

ViliusS replied:
@vy thank you for the detailed analysis. I have tested exception resolver with another exception type on Artemis and it looks light you might be right. The output with exception resolver during native Java's IllegalStateException:
![image](https://github.com/user-attachments/assets/d6916995-c75e-4c5a-9c25-5950e306c36e)

I have now changed JTL to use exception resolver as requested.

If there are no other issues with this PR I will then try to fix tests for timestamp fields.

@ppkarwasz I do have access to Artemis JIRA and will send PR to add full package name to them shortly.

ViliusS commented:
> @ViliusS,
> 
> Thanks for you contribution.
> 
> Did you test it on an actual Google Cloud instance? I am not a GCP user, but some of the changes you propose differ from the documented log format.

Yes I did. Here are some screenshots.

Exception caught because of the stack trace in message field:
![image](https://github.com/user-attachments/assets/1b5fc3e8-3f01-40b1-8353-a9c3bcafa3e3)
Expanded jsonPayload fields in the logging:
![image](https://github.com/user-attachments/assets/039fdd90-a23e-428a-bccd-959f16f0fa2f)
Proof about automatic timestamp mapping (previously this field was generated by Google and the one sent from Log4J lived in `jsonPayload.timestamp`:
![image](https://github.com/user-attachments/assets/10139b2a-1587-4a46-8c28-b5c7d7869182)

The same error caught in Error Reporting UI:
![image](https://github.com/user-attachments/assets/9e2ce7df-2415-4ecf-b415-c5b602228b8b)

ppkarwasz commented on log4j-layout-template-json/src/main/resources/GcpLayout.json:
```diff
@@ -37,8 +42,9 @@
-    "stringified": true
+    "$resolver": "pattern",
+    "pattern": "%uuid{TIME}",
+    "stackTraceEnabled": false
```
Fixed in https://github.com/apache/logging-log4j2/pull/3586/commits/b19491c451333464442416eff462bbbd5750cd00

vy replied:
> `counter` produces duplicate IDs for different thread of the same application.

@ViliusS, what do you mean by this? Each JTL layout instance has its own `AtomicLong` (or `AtomicBigInteger`) counter. If you end up seeing the same counter value, it sounds like a severe bug to me.

ViliusS replied:
@vy here are the GCP logs with counter type `insertId` for the ActiveMQ Artemis instance:

![image](https://github.com/user-attachments/assets/117ffcf3-4f38-4c3f-a9b3-fea97184452d)

The first green column is the `insertId`, the second column is thread name. The log up to the counter 40 is startup log. Then I have stopped Java app (Artemis instance) after half an hour and insertId started from 0 for shutdown log. Then it restarts at 0 second time, when I again start the app. Not sure if that's a bug in how Artemis used Log4J for logging or a bug in Log4J itself, but in any case, restarting from 0 after the application was restarted is not what this field supposed to represent.

This is how it looks like with time-based UUIDs:
![image](https://github.com/user-attachments/assets/83b70aa5-fd08-467e-a0bc-7ee649915aed)

And this is the look with the current patch, where insertIds are generated by GCP themselves:
![image](https://github.com/user-attachments/assets/b00da79a-a23e-4280-bbd1-808b91b39ab8)

vy replied:
@ViliusS, I have the inkling that those messages containing overlapping `insertId`'s are not originating from the same host. But all in all, indeed _"a counter unique per JTL instance"_ does not guarantee the uniqueness required by `insertId`. We are better off just omitting that field and letting it be auto-generated by GCP.

ppkarwasz commented:
Let's wait for @vy's review.

@ViliusS, in this repository we require signed commits, so you need to:

1. Add your GPG public key to [your GitHub Settings](https://github.com/settings/keys) if you didn't do that already.
2. Rebase the PR to resign all the commits.

vy commented:
One particular thing that is bothering me about this _fix_ is... it breaks backward compatibility. We can say we're fixing the schema, but for users who have already been using the _broken_ schema, we are gonna create trouble. We can

* Cross our fingers, release the fixed schema, and think of a solution if we ever happen to receive a complaint
* Release the new schema under a different name (e.g., `GcpLayout2.json`)

@ppkarwasz, thoughts?

@ViliusS, would you mind adding a changelog entry file (i.e., `src/changelog/.2.x.x/3586_fix_GcpLayout.xml`), please? (Mind the `entry.type` attribute and `entry > issue` element.)

ppkarwasz commented:
> One particular thing that is bothering me about this _fix_ is... it breaks backward compatibility. We can say we're fixing the schema, but for users who have already been using the _broken_ schema, we are gonna create trouble. We can
> 
> * Cross our fingers, release the fixed schema, and think of a solution if we ever happen to receive a complaint
> * Release the new schema under a different name (e.g., `GcpLayout2.json`)

I would call the new format `GoogleOpsAgent.json`, `FluentBit.json` or something similar.

However, I am not entirely sure that this template file belongs to Log4j. The only users that can profit from it are Google Ops Agent users, so we might consider moving the file to that project instead. See also https://github.com/GoogleCloudPlatform/ops-agent/issues/1920#issuecomment-2823869658

ViliusS commented:
> One particular thing that is bothering me about this _fix_ is... it breaks backward compatibility. We can say we're fixing the schema, but for users who have already been using the _broken_ schema, we are gonna create trouble. We can
> 
> * Cross our fingers, release the fixed schema, and think of a solution if we ever happen to receive a complaint
> * Release the new schema under a different name (e.g., `GcpLayout2.json`)
> 
> @ppkarwasz, thoughts?
> 
> @ViliusS, would you mind adding a changelog entry file (i.e., `src/changelog/.2.x.x/3586_fix_GcpLayout.xml`), please? (Mind the `entry.type` attribute and `entry > issue` element.)

Changelog added.

Regarding backward compatibility I see it this way. The proposed changes are not really fixes, they are changes/improvements. Old schema probably worked in some sense, but required a little bit tinkering for proper GCP field mapping on Google side. It is natural that schemas change and they need to be re-adjusted, there is no way to avoid that when dealing with 3rd party services. The real question is when backward compatibility and how it should be broken.
You are right that some users could be using old schema already, so let's try to list what's actually changed and the use cases.
1) The `timestamp` field users who just send the data from previous field directly to GCP not realizing that it is not mapped to real log timestamp. Nothing serious changes for them, new timestamp fields will be sent to GCP and automatically mapped to log timestamp. They will be missing previous `jsonPayload.timestamp` field on GCP side, but that should not break anything serious.
2) `Timestamp` field users who processed current field values with Ops Agent or any other tool to be able to map it to GCP timestamp. Such tools usually just ignore missing field, but again information is now sent to GCP's timestamp field automatically, so nothing is lost.
3) Removed `insertId` field. Nothing should depend on this field. It is for internal GCP usage and will now be generated automatically.
4) Renamed/Changed `exception` format and removed separate `class` and `message` subfields. Some could depend on these fields in Google Cloud Monitoring but would consider it as an edge case, since the exception was/is duplicated in main `message` field anyway and was/is caught by the Google Error Reporting automatically. There was/is not need to configure separate monitoring.
5) Renamed `thread` and `logger` fields. Some could depend on these fields in Google Cloud Monitoring but, again, I would consider it as an edge case.

Considering that new adjustments makes it a lot easier to send logs to GCP now without major processing in between, and if they are included in 2.25.0, I think it is fair.

@ppkarwasz I wouldn't call it `OpsAgent`. This format could be used to print directly to stdout which is then automatically sent to GCP on Google's Kubernetes Engine or serverless AppEngine services. I would not differentiate it from ECS or Gelf you already support.
I also don't think it belongs to Ops Agent repo. If, for example, I want to use two monitoring systems and both require different formats moving JSON template elsewhere means that I must use Ops Agent for one monitoring solution, and then use different agent for another solution. Having JSON templates in central repo makes things must easier and more transparent.

ppkarwasz commented:
> I wouldn't call it `OpsAgent`. This format could be used to print directly to stdout which is then automatically sent to GCP on Google's Kubernetes Engine or serverless AppEngine services. I would not differentiate it from ECS or Gelf you already support.

Looking at https://github.com/GoogleCloudPlatform/ops-agent/issues/1920#issuecomment-2769381753 there are currently three different ways to interpret structured logs in Google Cloud:

1. The [legacy `fluentd`-based agent](https://cloud.google.com/logging/docs/agent/logging/configuration#special-fields), which we can probably disregard.
2. The default [`out_stackdriver` configuration](https://github.com/fluent/fluent-bit-docs/blob/master/pipeline/outputs/stackdriver_special_fields.md).
3. The [special fields interpreted by `OpsAgent`](https://cloud.google.com/logging/docs/agent/ops-agent/configuration#special-fields).

So I think we should use a template name that suggests, which of these formats is supported, like `GoogleOpsAgent.json`. This way we can further specialize the format by:

- Replacing `severity` with `logging.googleapis.com/severity`.
- Removing the unsupported `logging.googleapis.com/trace_sampled` field.

**Note**: We can also add multiple templates, one for `stackdriver` and one for `OpsAgent`. What do you think? To properly maintain the new templates, for me it is important to know **exactly** which specification we need to follow.

ViliusS commented:
> > I wouldn't call it `OpsAgent`. This format could be used to print directly to stdout which is then automatically sent to GCP on Google's Kubernetes Engine or serverless AppEngine services. I would not differentiate it from ECS or Gelf you already support.
> 
> Looking at [GoogleCloudPlatform/ops-agent#1920 (comment)](https://github.com/GoogleCloudPlatform/ops-agent/issues/1920#issuecomment-2769381753) there are currently three different ways to interpret structured logs in Google Cloud:
> 
> 1. The [legacy `fluentd`-based agent](https://cloud.google.com/logging/docs/agent/logging/configuration#special-fields), which we can probably disregard.
> 2. The default [`out_stackdriver` configuration](https://github.com/fluent/fluent-bit-docs/blob/master/pipeline/outputs/stackdriver_special_fields.md).
> 3. The [special fields interpreted by `OpsAgent`](https://cloud.google.com/logging/docs/agent/ops-agent/configuration#special-fields).
> 
> So I think we should use a template name that suggests, which of these formats is supported, like `GoogleOpsAgent.json`. This way we can further specialize the format by:
> 
> * Replacing `severity` with `logging.googleapis.com/severity`.
> * Removing the unsupported `logging.googleapis.com/trace_sampled` field.
> 
> **Note**: We can also add multiple templates, one for `stackdriver` and one for `OpsAgent`. What do you think? To properly maintain the new templates, for me it is important to know **exactly** which specification we need to follow.

I thought about this earlier, but for me it looked like too few differences to support different templates. Also, because of those subtle differences, it could be complicated to document for the end users which template in which case needs to be used.

Given that OpsAgent is just one tool to process logs, and folks could be using any other log sending methods or processors, like for example native OTLP one, there is a question how many of these JSON templates we should have.

That's why I tried to come up with a format which is as close as possible natively supported by Google Cloud Looging itself. Sure you need to apply few tricks in every processor (in OpsAgent case for `severity` and `trace_sampled` fields), but in my mind these tools are there exactly for that reason, i.e. match log fields from source to destination, in case of differences.

Nevertheless, I will support any decision.

vy commented on src/changelog/.2.x.x/3586_improve_GcpLayout.xml:
```diff
@@ -0,0 +1,10 @@
+       type="changed">
+  <issue id="3586" link="https://github.com/apache/logging-log4j2/pull/3586"/>
+  <description format="asciidoc">
+    Adjust GcpLayout JSON template to support automatic timestamp recognition by the Google Cloud Logging. This also changes `exception`, `thread`, `logger` fields a bit, and removes `insertId` field.
```
```suggestion
    Update `GcpLayout.json` JSON Template Layout event template to support automatic timestamp recognition by the Google Cloud Logging. This also changes `exception`, `thread`, `logger` fields, and removes `insertId` field.
```

vy commented:
@ViliusS, @ppkarwasz, I think we better proceed with the current state of this PR – I've approved it.
